hpy/devel/src/runtime/structseq.c

```diff
@@ -0,0 +1,396 @@
+        Py_DECREF(seq);
+        goto error;
+    }
+
+    PyObject *item;
+    for (Py_ssize_t i = 0; i < nargs; i++) {
+        item = _h2py(args[i]);
+        if (!args_owned)
+            Py_INCREF(item);
+        PyStructSequence_SetItem(seq, i, item);
+    }
+    /* no need to call 'close_array' if 'args_owned' because items are stolen
+       from 'args' by 'PyStructSequence_SetItem' */
+    return _py2h(seq);
+
+type_error:
+    /* The error behavior seems not to be consistent. If anything goes
+       wrong, we will convert it to a TypeError here. */
+    PyErr_Clear();
+    PyErr_Format(PyExc_TypeError,
+                 "type '%s' does not look like a struct sequence type",
+                 tp->tp_name);
+
+error:
+    if (args_owned)
+        close_array(ctx, nargs, args);
+    return HPy_NULL;
+#endif
+}
+
+HPyAPI_HELPER HPy
+HPyStructSequence_New(HPyContext *ctx, HPy type, HPy_ssize_t nargs, HPy *args)
+{
+    return structseq_new(ctx, type, nargs, args, false);
+}
+
+HPyAPI_HELPER HPy
+HPyStructSequence_NewFromFormat(HPyContext *ctx, HPy type, const char *fmt, ...)
+{
+    va_list values;
+    HPy result;
+    va_start(values, fmt);
+    HPy_ssize_t size = count_single_items(ctx, fmt);
+    if (size < 0) {
+        result = HPy_NULL;
+    } else if (size == 0) {
+        result = HPyStructSequence_New(ctx, type, 0, NULL);
+    } else if (size == 1) {
+        int owned;
+        result = buildvalue_single(ctx, &fmt, &values, &owned);
+        if (!owned) {
+            result = HPy_Dup(ctx, result);
+        }
+    } else {
+        result = build_array(ctx, type, &fmt, &values, size);
+    }
+    va_end(values);
+    return result;
+
+}
+
+static const char ERROR_FMT[] = "bad format char '%c' in the format string passed to HPyStructSequence_NewFromFormat";
+
+static HPy_ssize_t
+count_single_items(HPyContext *ctx, const char *fmt)
+{
+    HPy_ssize_t i;
+    char c;
+    for (i = 0; (c = fmt[i]) != '\0'; i++) {
+        switch (c) {
+        case 'i':
+        case 'I':
+        case 'k':
+        case 'l':
+        case 'L':
+        case 'K':
+        case 's':
+        case 'O':
+        case 'S':
+        case 'N':
+        case 'f':
+        case 'd':
+            break;
+
+        default: {
+            /* the size of ERROR_FMT will be enough since we are just inserting
+               a single char */
+            char msg[sizeof(ERROR_FMT)];
+            snprintf(msg, sizeof(msg), ERROR_FMT, c);
+            HPyErr_SetString(ctx, ctx->h_SystemError, msg);
+            return -1;
+        }
+        }
+    }
+    return i;
+}
+
+static void
+close_array(HPyContext *ctx, HPy_ssize_t n, HPy *arr)
+{
+    for (HPy_ssize_t i=0; i < n; i++) {
+        HPy_Close(ctx, arr[i]);
+    }
+}
+
+static HPy
+build_array(HPyContext *ctx, HPy type, const char **fmt, va_list *values, HPy_ssize_t size)
+{
+#ifdef _WIN32
+    HPy *arr = _malloca(size);
+    HPy result;
+#else
+    HPy arr[size];
+#endif
```

Comment:
Why use `_malloca` on windows, it seems to complicate the code paths?

Reply:
Right but I was reading through [the manual](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/alloca?view=msvc-170) and it says:
```
Allocates memory on the stack. This function is deprecated because a more secure version is available; see _malloca.
```
The problem is: if `size` is unexpectedly large, then you would kill the stack (since max stack sizes are usually much smaller than heap sizes). `_malloca` is actually exactly that function I would also prefer on Linux and other systems. It decides whether to allocate on stack or heap depending on the size. But I haven't seen such a function.

IMO, the increased complexity isn't that much. We need to call `_freea` (like we would need to do if we'd used `malloc`). I actually plan to use `_malloca` everywhere instead of `_alloca`.

Reply:
using `_malloca` (or `_alloca`) only on windows does not solve the stack problem, since it does not cover uses on non-windows. If there is no generic solution, I think I would prefer not to create different code paths for different runtme libraries. If you insist on using this, you should guard it with `#ifdef _MSC_VER` to prevent using this with non-MSVC compilers on windows.

Reply:
Note that we also use plain `alloca` elsewhere and just assume the stack has enough space. Maybe we should have some thin abstraction (also for `strncpy`, `snprintf`, ...), but not necessarily in this PR.

Reply:
> Maybe we should have some thin abstraction (also for strncpy, snprintf, ...), but not necessarily in this PR.

Agreed. I was also thinking about that. Should be easy for `strncpy`, `snprint` and such but for `_malloca`, it's not that easy. That needs to be a macro but that should also be possible.

Anyway, didn't do that since I was concentrating on getting the APIs for Milestone 1 in. I would do such enhancements after that.

Reply:
> using _malloca (or _alloca) only on windows does not solve the stack problem, since it does not cover uses on non-windows. If there is no generic solution, I think I would prefer not to create different code paths for different runtme libraries. 

You are right, we should have the same behavior as far as possible.

> If you insist on using this, you should guard it with #ifdef _MSC_VER to prevent using this with non-MSVC compilers on windows.

No, I don't insist on that. I'll remove it and will come up with a proper solution as Stepan suggested.